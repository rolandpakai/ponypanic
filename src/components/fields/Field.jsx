class Field {
  constructor(idd, type, level, theme, data) {
    this.idd = idd;
    this.level = level;
    this.type = type;
    this.theme = theme;
    this.data = data;

    this.popover = false;
  }

  getImgStyleSrc() {
    return this.imgStyleSrc;
  }

  getPopoverContent() {
    return this.popoverContent;
  }

  hasPopover() {
    return this.popover;
  }
}

export default Field;
