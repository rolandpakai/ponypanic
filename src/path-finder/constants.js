const Algorithms = Object.freeze({
  'A*': 'astar',
  'Dijkstra': 'dijkstra',
  'BFS': 'bfs'
});

const Cell = Object.freeze({
 'visited': -1,
 'empty': 0,
 'start': 1,
 'obstacle': 2,
 'end': 3,
 'path': 4,
});

const Heuristic = Object.freeze({
  'Manhattan': 0,
  'Euclidean': 1
});

if (typeof module !== 'undefined' && module.exports) {
  module.exports = exports = {Algorithms, Cell, Heuristic};
}