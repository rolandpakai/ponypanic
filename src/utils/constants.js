export const PLAYER_TOKEN = process.env.REACT_APP_PLAYER_TOKEN; 
// "760_I1ljZlZXWmUsSH0sVzNvajVmQH4yc11oUiQpPmNoLlVzUCQiSkh+aEY=";

export const TREASURE_COUNT = 9;
export const MAP_COUNT = 11;
export const IMG_BIG_SIZE = '64px';
export const IMG_SMALL_SIZE = '32px';
export const DEFAULT_THEME = 'pony';
export const PATH_REGEX = /(\d*[UDRL]|[UDRL])/g;

export const THEMES_SELECT_OPTIONS = [
  { value: 'pony', label: 'Pony' },
  { value: 'zelda', label: 'Zelda' },
]

export const MAP_STATUS = {
  CREATED: 'CREATED',
  PLAYING: 'PLAYING',
  WON: 'WON',
  LOST: 'LOST',
}

export const HERO_ACTION = {
  NOTHING: 'NOTHING',
  USE_SHIELD: 'USE_SHIELD',

  MOVE_LEFT: 'MOVE_LEFT',
  MOVE_RIGHT: 'MOVE_RIGHT',
  MOVE_UP: 'MOVE_UP',
  MOVE_DOWN: 'MOVE_DOWN',

  KICK_LEFT: 'KICK_LEFT',
  KICK_RIGHT: 'KICK_RIGHT',
  KICK_UP: 'KICK_UP',
  KICK_DOWN: 'KICK_DOWN',
}

export const HERO_STATE = {
  STAND: 'stand', 
  SHILD: 'shild',
  ATTACK: 'attack',
}

export const HERO_MOVE = {
  LEFT: 'left', 
  SHILD: 'right',
  UP: 'up',
  DOWN: 'down',
}

export const FIELD_TYPE = {
  HERO: 'HERO',
  TREASURE: 'TREASURE',
  ENEMY: 'ENEMY',
  BULLET: 'BULLET',
  OBSTACLE: 'OBSTACLE',
  FLOOR: 'FLOOR',
}