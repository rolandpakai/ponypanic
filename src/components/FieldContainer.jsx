import React, { useContext, useState } from "react";
import Box from "@mui/material/Box";
import Popover from "@mui/material/Popover";

import Table from "@mui/material/Table";
import TableCell, { tableCellClasses } from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";

import { ThemeContext } from "../contexts/ThemeContext";
import { FIELD_TYPE, THEME_MAP_COUNT } from "../utils/constants";

const FieldContainer = (props) => {
  const [anchorEl, setAnchorEl] = useState();
  const { theme } = useContext(ThemeContext);
  const { idd, size, type, level, field } = { ...props };

  const open = Boolean(anchorEl);

  const fieldLevel = level > THEME_MAP_COUNT ? THEME_MAP_COUNT : level;

  const handlePopoverOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handlePopoverClose = () => {
    setAnchorEl(null);
  };

  const imgStyle = {
    width: size,
    height: size,
    backgroundRepeat: "no-repeat",
    margin: "auto",
    src: "",
  };

  return (
    <>
      <Box
        id={`field-${idd}`}
        aria-owns={
          field.hasPopover() && open ? "mouse-over-popover" : undefined
        }
        aria-haspopup="true"
        onMouseEnter={field.hasPopover() ? handlePopoverOpen : undefined}
        onMouseLeave={field.hasPopover() ? handlePopoverClose : undefined}
        sx={{
          width: size,
          height: size,
          display: "flex",
          backgroundSize: size,
          backgroundRepeat: "no-repeat",
          backgroundImage: `url('./themes/${theme}/maps/map-${fieldLevel}/floor.png')`,
        }}
      >
        {type !== FIELD_TYPE.FLOOR && (
          <img
            alt={type}
            src={`./themes/${theme}/${field.getImgSrc()}`}
            style={imgStyle}
          />
        )}
      </Box>
      {field.hasPopover() && (
        <Popover
          anchorEl={anchorEl}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          disableRestoreFocus
          id="mouse-over-popover"
          onClose={handlePopoverClose}
          open={open}
          sx={{
            pointerEvents: "none",
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "left",
          }}
        >
          <TableContainer>
            <Table
              aria-label="simple table"
              size="small"
              sx={{
                [`& .${tableCellClasses.root}`]: {
                  borderBottom: "none",
                },
              }}
            >
              <TableHead>
                <TableRow>
                  <TableCell align="center" colSpan={2}>
                    {type}
                  </TableCell>
                </TableRow>
              </TableHead>
              {field.getPopoverContent()}
            </Table>
          </TableContainer>
        </Popover>
      )}
    </>
  );
};

export default FieldContainer;
