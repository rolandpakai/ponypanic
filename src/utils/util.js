import { PATH_REGEX, HERO_ACTION, IMG_BIG_SIZE, IMG_SMALL_SIZE } from '../utils/constants';

export const randomInteger = (min, max) => {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

export const timeoutPromise = (timeout) => {
  return new Promise(resolve => {
      setTimeout(resolve, timeout);
  });
}

export const xyTOij = (x, y, height) => {
  return { i: x, j: height - y - 1 }
}

export const arrayToMap = (array, type, turn) => {
  const map = {};

  array.forEach((el) => {
    const idd = `${el.position.x}-${el.position.y}`;
    el.type = type;

    if(turn && turn.heroId && turn.heroId === el.id) {
      el.action = turn.action;
    }

    map[idd] = el;
  });

  return map;
}

export const mapToArray = (map) => {
  const array = [];
  
  for (const x in map) {
    if (map.hasOwnProperty(x)) {
      map[x].map((y) => array.push({position: {x, y}}))
    }
  }

  return array;
}

export const addEntityToMap = (map, array, type) => {
  array.forEach((el) => {
    const idd = `${el.position.x}-${el.position.y}`;
    el.type = type;
    map[idd] = el;
  })
}

export const getImageSize = (mapSize) => {
  return mapSize > 10 ?  IMG_SMALL_SIZE : IMG_BIG_SIZE;
}

export const getHeroAction = (direction) => {
  let action = HERO_ACTION.NOTHING;

  switch (direction) {
    case 'U': {action = HERO_ACTION.MOVE_UP; break;}
    case 'D': {action = HERO_ACTION.MOVE_DOWN; break;}
    case 'L': {action = HERO_ACTION.MOVE_LEFT; break;}
    case 'R': {action = HERO_ACTION.MOVE_RIGHT; break;}
    default: action = HERO_ACTION.NOTHING;
  }

  return action;
}

export const validateHeroAction = (heroAction) => {
  let action = HERO_ACTION.MOVE_RIGHT.toLowerCase();

  if(heroAction && HERO_ACTION[heroAction]) {
    action = HERO_ACTION[heroAction].toLowerCase();
  }

  return action;
}

export const calcDirection = (paths) => {

  const pathLengths = paths.map((path) => {
    const ways = path.match(PATH_REGEX);
    
    const length = ways.reduce((acc, way) => {
      let stepLength = 0;
      const match = way.match(/^\d+/);
      
      if(match) {
        stepLength = parseInt(match[0], 10);
      } else {
        stepLength = 1;
      }
      
      return acc + stepLength;
    }, 0);
    
    return length;
  });
  
  const minLength = Math.min(...pathLengths);
  const minLengthIndex = pathLengths.indexOf(minLength);
  const minPath = paths[minLengthIndex];
  const firstStepInMinPath = minPath.match(PATH_REGEX)[0]
  const direction = firstStepInMinPath.charAt(firstStepInMinPath.length-1);

  return direction;
}

export const getDirections = (pathsList) => {
  let directions = [];

  if(pathsList.length > 0){
    directions = pathsList.map((paths) => {
      return calcDirection(paths)
    })
  }

  return directions;
}
