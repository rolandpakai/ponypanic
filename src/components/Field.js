import { useMemo, useContext } from "react";

import { ThemeContext } from '../contexts/ThemeContext';
import { randomInteger, validateHeroAction } from '../utils/util';
import { FIELD_TYPE, TREASURE_COUNT, MAP_COUNT, ENEMY_COUNT } from '../utils/constants';

const Field = ( props ) => {
  const { theme } = useContext(ThemeContext);
  const { id, collectedByHeroId, health, name, playerId, position, size, score, action, type, level } = {...props};

  let randomTreasure = useMemo(() => randomInteger(1, TREASURE_COUNT), [level]);
  let randomEnemy = useMemo(() => randomInteger(1, ENEMY_COUNT), [level]);

  const mapLevel = level > MAP_COUNT ? MAP_COUNT : level;

  const divStyle = {
    width: size,
    height: size,
    backgroundSize: size,
    backgroundRepeat: 'no-repeat',
    backgroundImage: `url('./themes/${theme}/maps/map-${mapLevel}/floor.png')`,
  }

  let imgStyle = {
    width: size,
    height: size,
    backgroundRepeat: 'no-repeat',
    src: '',
  };

  switch (type) {
    case FIELD_TYPE.HERO: {
      const heroAction = validateHeroAction(action);
      imgStyle.src = `./themes/${theme}/heroes/${heroAction}.png`;
      break;
    } 
    case FIELD_TYPE.ENEMY: {
      imgStyle.src = `./themes/${theme}/enemies/enemy-${randomEnemy}.png`;
      break;
    }
    case FIELD_TYPE.BULLET: {
      imgStyle.src = `./themes/${theme}/bullets/bullet.png`;
      break;
    }
    case FIELD_TYPE.TREASURE: {
      imgStyle.src = `./themes/${theme}/treasures/treasure-${randomTreasure}.png`;
      break;
    }
    case FIELD_TYPE.OBSTACLE: {
      imgStyle.src = `./themes/${theme}/maps/map-${mapLevel}/block.png`;
      break;
    }
    default: imgStyle.backgroundImage = '';
  }

  return (
      <div style={divStyle}>
        { type !== FIELD_TYPE.FLOOR &&
          <img 
            alt={type}
            className=''
            src={imgStyle.src}
            style={imgStyle}
          />
        }
      </div>
  )
}

export default Field;