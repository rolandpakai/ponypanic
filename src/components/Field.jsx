import React, { Fragment, useContext, useMemo, useState } from "react";
import Box from "@mui/material/Box";
import Popover from "@mui/material/Popover";

import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell, { tableCellClasses } from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";

import { ThemeContext } from "../contexts/ThemeContext";
import { randomInteger, validateHeroAction } from "../utils/util";
import { FIELD_TYPE, MAP_COUNT, TREASURE_COUNT } from "../utils/constants";

const Field = (props) => {
  const [anchorEl, setAnchorEl] = useState();
  const { theme } = useContext(ThemeContext);
  const { idd, size, action, type, level, data } = { ...props };

  const open = Boolean(anchorEl);

  let hasPopOver = false;

  const randomTreasure = useMemo(() => randomInteger(1, TREASURE_COUNT), []);
  const mapLevel = level > MAP_COUNT ? MAP_COUNT : level;

  const handlePopoverOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handlePopoverClose = () => {
    setAnchorEl(null);
  };

  const imgStyle = {
    width: size,
    height: size,
    backgroundRepeat: "no-repeat",
    margin: "auto",
    src: "",
  };

  let popoverContent;

  switch (type) {
    case FIELD_TYPE.HERO: {
      const { playerId, health, score } = { ...data };
      const heroAction = validateHeroAction(action).toLowerCase();

      imgStyle.src = `./themes/${theme}/heroes/${heroAction}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-playerId`}>
            <TableCell align="left">Id</TableCell>
            <TableCell align="right">{playerId}</TableCell>
          </TableRow>
          <TableRow key={`${idd}-health`}>
            <TableCell align="left">Health</TableCell>
            <TableCell align="right">
              {health.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-score`}>
            <TableCell align="left">Score</TableCell>
            <TableCell align="right">{score}</TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.ENEMY:
    case FIELD_TYPE.ENEMY_BULLET: {
      const {
        id,
        moveProbability,
        shootProbability,
        onTouchDamage,
        bulletDamage,
        health,
      } = { ...data };

      imgStyle.src = `./themes/${theme}/enemies/${type.toLowerCase()}-${mapLevel}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-id`}>
            <TableCell align="left">Id</TableCell>
            <TableCell align="right">{id}</TableCell>
          </TableRow>
          <TableRow key={`${idd}-health`}>
            <TableCell align="left">Health</TableCell>
            <TableCell align="right">
              {health.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-moveProbability`}>
            <TableCell align="left">Move Probability</TableCell>
            <TableCell align="right">
              {moveProbability.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-shootProbability`}>
            <TableCell align="left">Shoot Probability</TableCell>
            <TableCell align="right">
              {shootProbability.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-bulletDamage`}>
            <TableCell align="left">Bullet Damage</TableCell>
            <TableCell align="right">
              {bulletDamage.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-onTouchDamage`}>
            <TableCell align="left">Touch Damage</TableCell>
            <TableCell align="right">
              {onTouchDamage.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.BULLET: {
      // const { id, damage, direction, shotByEnemyId } = { ...data };

      imgStyle.src = `./themes/${theme}/bullets/bullet.png`;
      imgStyle.width = "auto";
      imgStyle.height = "auto";
      break;
    }
    case FIELD_TYPE.TREASURE: {
      const { name } = { ...data };

      imgStyle.src = `./themes/${theme}/treasures/treasure-${randomTreasure}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-name`}>
            <TableCell align="center">{name}</TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.OBSTACLE: {
      imgStyle.src = `./themes/${theme}/maps/map-${mapLevel}/block.png`;
      break;
    }
    default:
      imgStyle.backgroundImage = "";
  }

  return (
    <>
      <Box
        aria-owns={hasPopOver && open ? "mouse-over-popover" : undefined}
        aria-haspopup="true"
        onMouseEnter={hasPopOver ? handlePopoverOpen : undefined}
        onMouseLeave={hasPopOver ? handlePopoverClose : undefined}
        sx={{
          width: size,
          height: size,
          display: "flex",
          backgroundSize: size,
          backgroundRepeat: "no-repeat",
          backgroundImage: `url('./themes/${theme}/maps/map-${mapLevel}/floor.png')`,
        }}
      >
        {type !== FIELD_TYPE.FLOOR && (
          <img alt={type} className="" src={imgStyle.src} style={imgStyle} />
        )}
      </Box>
      {hasPopOver && (
        <Popover
          anchorEl={anchorEl}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          disableRestoreFocus
          id="mouse-over-popover"
          onClose={handlePopoverClose}
          open={open}
          sx={{
            pointerEvents: "none",
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "left",
          }}
        >
          <TableContainer>
            <Table
              aria-label="simple table"
              size="small"
              sx={{
                [`& .${tableCellClasses.root}`]: {
                  borderBottom: "none",
                },
              }}
            >
              <TableHead>
                <TableRow>
                  <TableCell align="center" colSpan={2}>
                    {type}
                  </TableCell>
                </TableRow>
              </TableHead>
              {popoverContent}
            </Table>
          </TableContainer>
        </Popover>
      )}
    </>
  );
};

export default Field;
