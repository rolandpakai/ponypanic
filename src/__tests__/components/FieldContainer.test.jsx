import React from "react";
import { render } from "@testing-library/react";
import FieldContainer from "../../components/FieldContainer";
import Field from "../../components/fields/Field";
import { FIELD_TYPE, IMG_BIG_SIZE } from "../../utils/constants";

describe("FieldContainer Component", () => {
  test("FieldContainer renders", () => {
    const { container } = render(
      <FieldContainer
        idd="idd"
        size={IMG_BIG_SIZE}
        type={FIELD_TYPE.FLOOR}
        level={1}
        field={new Field("idd", FIELD_TYPE.FLOOR, 1, {})}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
