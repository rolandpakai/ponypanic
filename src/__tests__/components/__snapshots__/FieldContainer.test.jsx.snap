// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldContainer Component FieldContainer renders 1`] = `
<div>
  <div
    aria-haspopup="true"
    class="MuiBox-root css-1sdm1q0"
    id="field-idd"
  />
</div>
`;
